src/open-uri.c

```diff
@@ -202,18 +138,130 @@ launch_application_with_uri (const char *choice_id,
+update_permissions_store (const char *app_id,
+                          const char *content_type,
+                          const char *chosen_id)
 {
-  Request *request = request_from_invocation (invocation);
+  g_autoptr(GError) error = NULL;
+  g_autofree char *latest_chosen_id = NULL;
+  gint latest_chosen_count = 0;
+  g_auto(GStrv) in_permissions = NULL;
+
+  if (get_latest_choice_info (app_id, content_type, &latest_chosen_id, &latest_chosen_count) &&
+      (g_strcmp0 (chosen_id, latest_chosen_id) == 0))
+    {
+      /* same app chosen once again: update the counter */
+      if (latest_chosen_count >= USE_DEFAULT_APP_THRESHOLD)
+        latest_chosen_count = USE_DEFAULT_APP_THRESHOLD;
+      else
+        latest_chosen_count++;
+    }
+  else
+    {
+      /* latest_chosen_id is heap-allocated */
+      latest_chosen_id = g_strdup (chosen_id);
+      latest_chosen_count = 0;
+    }
+
+  in_permissions = (GStrv) g_new0 (char *, 2);
+  in_permissions[0] = g_strdup_printf ("%s:%u", latest_chosen_id, latest_chosen_count);
+
+  if (!xdp_impl_permission_store_call_set_permission_sync (permission_store_impl,
+                                                           TABLE_NAME,
+                                                           TRUE,
+                                                           content_type,
+                                                           app_id,
+                                                           (const char * const*) in_permissions,
+                                                           NULL,
+                                                           &error))
+    {
+      g_warning ("Error updating permission store: %s\n", error->message);
+      g_clear_error (&error);
+    }
+}
+
+static void
+send_response_in_thread_func (GTask        *task,
+                              gpointer      source_object,
+                              gpointer      task_data,
+                              GCancellable *cancellable)
+{
+  Request *request = (Request *)task_data;
+  guint response;
+  GVariant *options;
+  const char *uri;
+  const char *parent_window;
+  const char *content_type;
+  const char *choice;
+
+
+  REQUEST_AUTOLOCK (request);
+
+  response = GPOINTER_TO_INT (g_object_get_data (G_OBJECT (request), "response"));
+  options = (GVariant *)g_object_get_data (G_OBJECT (request), "options");
+
+  if (response != 0)
+    goto out;
+
+  g_variant_lookup (options, "&s", "choice", &choice);
+
+  uri = g_object_get_data (G_OBJECT (request), "uri");
+  parent_window = g_object_get_data (G_OBJECT (request), "parent-window");
+  content_type = g_object_get_data (G_OBJECT (request), "content-type");
+
+  launch_application_with_uri (choice, uri, parent_window);
+  update_permissions_store (request->app_id, content_type, choice);
+
+out:
+  if (request->exported)
+    {
+      xdp_request_emit_response (XDP_REQUEST (request), response, options);
+      request_unexport (request);
+    }
```

Comment:
In fact, a lot of things are still repeated, althought less so than before. I wonder if we can generalize things so that e.g. the request creation and locking is automatic.